src/include/sof/drivers/interrupt.h

```diff
@@ -12,12 +12,28 @@
 #include <platform/drivers/interrupt.h>
 #include <sof/lib/cpu.h>
 #include <sof/list.h>
-#include <sof/spinlock.h>
 #include <sof/trace/trace.h>
 #include <user/trace.h>
 #include <stdbool.h>
 #include <stdint.h>
 
+/* For spinlock_t we cannot directly include <sof/spinlock.h> anymore. But
+ * we only need this type which is defined in <arch/spinlock.h>. But in
+ * order to get to it we need to pretend we are including
+ * <sof/spinlock.h> to avoid the compile error.
+ *
+ * Also, if that file is already included we don't need further
+ * includes.
+ *
+ * Solves a circular dependency.
+ */
+
+#ifndef __SOF_SPINLOCK_H__
+#define __SOF_SPINLOCK_H__
+#include <arch/spinlock.h>
+#undef __SOF_SPINLOCK_H__
+#endif
+
```

Comment:
I need to know your opinion on this one in particular, I couldn't find a better solution to solve this circular dependency of headers so here is what I have done.

Reply:
Do you know what headers are involved here, it may be easier to fix anothe header or to move some definintions around to solve it.

Reply:
Specifically it's <sof/spinlock.h> and <sof/drivers/interrupt.h> which currently include each other. In the current, macro variant it doesn't matter what order the macros are defined, but in the new variant it does. This is the least invasive of the circular dependency removal that I could find (<sof/drivers/interrupt.h> doesn't need the full <sof/spinlock.h> header but only the type spinlock_t which is in <arch/spinlock.h>)

Reply:
We should log this a bug. It's just going to be more pain later on.

Reply:
That undef definitely cannot be there.
Easiest would be to also typedef spintlock_t in interrupt, but it's duplication, so meh, but works.
One of the more elegant ways may be putting common part in another header and include it in both headers, but it's weird to do it just for typedef, so maybe it should go to some common header for atomic-related types.

Reply:
I'd like to have circ dependency issue fixed first.

Reply:
@jajanusz If we extract the `spinlock_t` type in a separate header we would fix this particular circular dependency issue. If you wish I could try to do an analysis to find if there are other instances of this issue.

Reply:
@jajanusz Extracted it into `<sof/spinlock_t.h>`, please re-review.